Panic on startup if checkpoint directory is empty

If the checkpoint directory is present but has no files you get the following panic:

```
level=info ts=2020-03-13T08:53:02.071Z caller=head.go:584 component=tsdb msg="replaying WAL, this may take awhile"
panic: runtime error: index out of range [0] with length 0

goroutine 108 [running]:
github.com/prometheus/prometheus/tsdb/wal.NewSegmentBufReader(...)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:830
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsRangeReader(0xc00009baf8, 0x1, 0x1, 0x1a, 0xb77, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:813 +0x760
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsReader(...)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/wal/wal.go:785
github.com/prometheus/prometheus/tsdb.(*Head).Init(0xc0000ee780, 0x8000000000000000, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/head.go:592 +0xbc4
github.com/prometheus/prometheus/tsdb.Open(0x24dec1b, 0x5, 0x2c64440, 0xc00060b170, 0x2c9f6a0, 0xc000144460, 0xc0000a3900, 0x2, 0xc0005fb000, 0xc)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/tsdb/db.go:567 +0x5f9
github.com/prometheus/prometheus/storage/tsdb.Open(0x24dec1b, 0x5, 0x2c64440, 0xc00060b170, 0x2c9f6a0, 0xc000144460, 0xc00014a1e0, 0x0, 0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/storage/tsdb/tsdb.go:192 +0x290
main.main.func19(0x0, 0x0)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:654 +0x31a
github.com/oklog/run.(*Group).Run.func1(0xc0000ca840, 0xc000126640, 0xc00057c000)
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
        /home/bbrazil/gosrc/src/github.com/prometheus/prometheus/vendor/github.com/oklog/run/group.go:37 +0xbb
```

We should be resilient to this, and also check what can cause this to occur.

I have exactly this same issue at the moment, and this is the only reference to it I could find

```
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:332 msg="Starting Prometheus" version="(version=2.14.0, branch=HEAD, revision=edeb7a44cbf745f1d8be4ea6f215e79e651bfe19)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:333 build_context="(go=go1.13.4, user=root@df2327081015, date=20191111-14:27:12)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:334 host_details="(Linux 4.15.0-88-generic #88-Ubuntu SMP Tue Feb 11 20:11:34 UTC 2020 x86_64 prom-net-dp-6bff67679c-hckq9 (none))"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:335 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2020-06-30T08:58:25.887Z caller=main.go:336 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-06-30T08:58:31.929Z caller=main.go:657 msg="Starting TSDB ..."
level=info ts=2020-06-30T08:58:31.929Z caller=web.go:496 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-06-30T08:58:32.252Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593403200000 maxt=1593410400000 ulid=01EBZE07WDE3F659NEPGHKVKKS
level=info ts=2020-06-30T08:58:32.256Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593410400000 maxt=1593417600000 ulid=01EBZMVZ4D451M5YTYWGPGGAEJ
level=info ts=2020-06-30T08:58:32.261Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593417600000 maxt=1593424800000 ulid=01EBZVQQYKDNBHT62EJY2T35AW
level=info ts=2020-06-30T08:58:32.265Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593424800000 maxt=1593432000000 ulid=01EC02KDMC866SKR14SE1QAEF5
level=info ts=2020-06-30T08:58:32.269Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593432000000 maxt=1593439200000 ulid=01EC09F4WCNAR75Q6FWE78N82A
level=info ts=2020-06-30T08:58:32.273Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593439200000 maxt=1593446400000 ulid=01EC0GAW4CZ92B8FS17KQNTGW3
level=info ts=2020-06-30T08:58:32.279Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593446400000 maxt=1593453600000 ulid=01EC0Q6KCDW48716V3DC7HGFB3
level=info ts=2020-06-30T08:58:32.284Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593453600000 maxt=1593460800000 ulid=01EC0Y2AMDM4F6NCRM39FQGTTZ
level=info ts=2020-06-30T08:58:32.288Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593460800000 maxt=1593468000000 ulid=01EC14Y1WCSVZSTZC5TSZTEQD1
level=info ts=2020-06-30T08:58:32.292Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593468000000 maxt=1593475200000 ulid=01EC1BSS4CSV3PRR8Z8F7VSXZW
level=info ts=2020-06-30T08:58:32.296Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593475200000 maxt=1593482400000 ulid=01EC1JNGCC8TCBXV0P77RCNDX5
level=info ts=2020-06-30T08:58:32.300Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593482400000 maxt=1593489600000 ulid=01EC1SH7MCB0KQ60MJ28KRX49C
level=info ts=2020-06-30T08:58:32.305Z caller=repair.go:59 component=tsdb msg="found healthy block" mint=1593489600000 maxt=1593496800000 ulid=01EC20CYWCEE03TYWXA6522AE6
level=info ts=2020-06-30T08:58:38.638Z caller=head.go:535 component=tsdb msg="replaying WAL, this may take awhile"
panic: runtime error: index out of range [0] with length 0

goroutine 192 [running]:
github.com/prometheus/prometheus/tsdb/wal.NewSegmentBufReader(...)
	/app/tsdb/wal/wal.go:830
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsRangeReader(0xc000d13af8, 0x1, 0x1, 0x21, 0x912, 0x0, 0x0)
	/app/tsdb/wal/wal.go:813 +0x760
github.com/prometheus/prometheus/tsdb/wal.NewSegmentsReader(...)
	/app/tsdb/wal/wal.go:785
github.com/prometheus/prometheus/tsdb.(*Head).Init(0xc000174000, 0x17303cf8b00, 0x0, 0x0)
	/app/tsdb/head.go:543 +0xbc4
github.com/prometheus/prometheus/tsdb.Open(0x7ffc15840407, 0xb, 0x2b4f400, 0xc00077c330, 0x2b8a220, 0xc0001cc460, 0xc0001b4b00, 0x2, 0xc00079a100, 0xc)
	/app/tsdb/db.go:562 +0x5f9
github.com/prometheus/prometheus/storage/tsdb.Open(0x7ffc15840407, 0xb, 0x2b4f400, 0xc00077c330, 0x2b8a220, 0xc0001cc460, 0xc0001d26e0, 0x0, 0x0, 0x0)
	/app/storage/tsdb/tsdb.go:194 +0x290
main.main.func19(0x0, 0x0)
	/app/cmd/prometheus/main.go:663 +0x31a
github.com/oklog/run.(*Group).Run.func1(0xc00019e420, 0xc0001b4a80, 0xc00077c210)
	/app/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
	/app/vendor/github.com/oklog/run/group.go:37 +0xbb
```

```
# ls -latr prometheus-data/
total 50
-rw-r--r--  1 root root     0 Feb 18 16:35 lock
drwxr-xr-x  3 root root  4096 Jun 17 21:16 thanos
drwxr-x---  3 root root  4096 Jun 24 11:06 ..
drwxr-xr-x  3 root root    64 Jun 29 09:01 01EBZMVZ25Y41BZ3D2FQ8K69QT
drwxr-xr-x  3 root root    64 Jun 29 11:01 01EBZVQP9T7TQHJMCBFF1MHSVF
drwxr-xr-x  3 root root    64 Jun 29 13:00 01EC02KDBYXNVHN0EMDA6HWYJW
drwxr-xr-x  3 root root    64 Jun 29 15:01 01EC09F4PX5TXW3HVHDRTW0Q9V
drwxr-xr-x  3 root root    64 Jun 29 17:01 01EC0GAVVT22F3NJ6HTS0H6PNF
drwxr-xr-x  3 root root    64 Jun 29 19:01 01EC0Q6K3YXKH3HK1Y1C6MGKE1
drwxr-xr-x  3 root root    64 Jun 29 21:01 01EC0Y2AEXZKDQH0285NRD52XD
drwxr-xr-x  3 root root    64 Jun 29 23:01 01EC14Y1M2HCGQRGH0KKZK0SSF
drwxr-xr-x  3 root root  4096 Jun 30 01:01 01EC1BSRYZH056R7P0K814DYEE
drwxr-xr-x  3 root root    64 Jun 30 03:01 01EC1JNG9RJZF43RV1M9ZREG6Z
drwxr-xr-x  3 root root    64 Jun 30 05:01 01EC1SH7F2FPM0M8NYP9D7DMW4
drwxr-xr-x  3 root root  4096 Jun 30 07:01 01EC20CYPT0WAWP099D878JMXN
drwxr-xr-x  3 root root  4096 Jun 30 09:01 01EC278P28EYCH1F7X4151B6EZ
drwxr-xr-x  3 root root    90 Jun 30 09:04 wal
-rw-r--r--  1 root root 20001 Jun 30 09:06 queries.active
-rw-r--r--  1 root root    68 Jun 30 09:06 thanos.shipper.json
drwxr-xr-x 18 root root  4096 Jun 30 09:06 .
```

Let me know if I can provide any useful logs.

--- a/tsdb/wal/checkpoint.go
+++ b/tsdb/wal/checkpoint.go
@@ -245,6 +245,20 @@ func Checkpoint(w *WAL, from, to int, keep func(id uint64) bool, mint int64) (*C
 	if err := cp.Close(); err != nil {
 		return nil, errors.Wrap(err, "close checkpoint")
 	}
+
+	// Sync temporary directory before rename.
+	df, err := fileutil.OpenDir(cpdirtmp)
+	if err != nil {
+		return nil, errors.Wrap(err, "open temporary checkpoint directory")
+	}
+	if err := df.Sync(); err != nil {
+		df.Close()
+		return nil, errors.Wrap(err, "sync temporary checkpoint directory")
+	}
+	if err = df.Close(); err != nil {
+		return nil, errors.Wrap(err, "close temporary checkpoint directory")
+	}
+
 	if err := fileutil.Replace(cpdirtmp, cpdir); err != nil {
 		return nil, errors.Wrap(err, "rename checkpoint directory")
 	}

